src/Sparrow/LowMemory/LowMemoryNotification.cs

```diff
@@ -0,0 +1,199 @@
+            }
+
+            inactiveHandlers.ForEach(x => _lowMemoryHandlers.TryRemove(x));
+        }
+
+        public void RegisterLowMemoryHandler(ILowMemoryHandler handler)
+        {
+            _lowMemoryHandlers.Add(new WeakReference<ILowMemoryHandler>(handler));
+        }
+
+        public static LowMemoryNotification Instance { get; private set; }
+        public bool LowMemoryState { get; set; }
+
+        public static void Initialize(CancellationToken shutdownNotification, long lowMemoryThreshold, double physicalRatioForLowMemDetection)
+        {
+            Instance = new LowMemoryNotification(shutdownNotification, lowMemoryThreshold, physicalRatioForLowMemDetection);
+        }
+
+        private readonly long _lowMemoryThreshold;
+        private readonly ManualResetEvent _simulatedLowMemory = new ManualResetEvent(false);
+        private readonly ManualResetEvent _shutdownRequested = new ManualResetEvent(false);
+        private readonly ManualResetEvent _warnAllocation = new ManualResetEvent(false);
+        private readonly double _physicalRatioForLowMemDetection;
+
+        public LowMemoryNotification(CancellationToken shutdownNotification, long lowMemoryThreshold, double physicalRatioForLowMemDetection)
+        {
+            _logger = LoggingSource.Instance.GetLogger<LowMemoryNotification>("Server");
+
+            shutdownNotification.Register(() => _shutdownRequested.Set());
+            _lowMemoryThreshold = lowMemoryThreshold;
+            _physicalRatioForLowMemDetection = physicalRatioForLowMemDetection;
+            new Thread(MonitorMemoryUsage)
+            {
+                IsBackground = true,
+                Name = "Low memory notification thread"
+            }.Start();
+        }
+
+        private void MonitorMemoryUsage()
+        {
+            int clearInactiveHandlersCounter = 0;
+            var memoryAvailableHandles = new WaitHandle[] { _simulatedLowMemory, _shutdownRequested };
+            var paranoidModeHandles = new WaitHandle[] { _simulatedLowMemory, _shutdownRequested, _warnAllocation };
+            var timeout = 5 * 1000;
+            long totalUnmanagedAllocations = 0;
+            while (true)
+            {
+                var handles = MemoryInformation.GetMemoryInfo().AvailableMemory.GetValue(SizeUnit.Bytes) < _lowMemoryThreshold * 2 ?
+                    paranoidModeHandles :
+                    memoryAvailableHandles;
+                switch (WaitHandle.WaitAny(handles, timeout))
+                {
+                    case WaitHandle.WaitTimeout:
+                        if (++clearInactiveHandlersCounter > 60) // 5 minutes == WaitAny 5 Secs * 60
+                        {
+                            clearInactiveHandlersCounter = 0;
+                            ClearInactiveHandlers();
+                        }
+                        var memInfo = MemoryInformation.GetMemoryInfo();
+                        foreach (var stats in NativeMemory.ThreadAllocations.Values
```

Comment:
This is doing a lot of allocations, and in potentially memory constrained situation.